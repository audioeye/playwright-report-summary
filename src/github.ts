import { getOctokit } from '@actions/github'
import { components } from "@octokit/openapi-types"

type IssueComment = components["schemas"]["issue-comment"];
type PullRequestReview = components["schemas"]["pull-request-review"];
type Octokit = ReturnType<typeof getOctokit>

export async function getIssueComments(octokit: Octokit, { owner, repo, issue_number }: { owner: string, repo: string, issue_number: number }): Promise<IssueComment[]> {
	const params = { owner, repo, issue_number }
	const { data: comments } = await octokit.rest.issues.listComments(params)
	return comments
}

export async function updateIssueComment(octokit: Octokit, { owner, repo, comment_id, body }: { owner: string, repo: string, comment_id: number, body: string }): Promise<void> {
	const params = { owner, repo, comment_id, body }
	await octokit.rest.issues.updateComment(params)
}

export async function createIssueComment(octokit: Octokit, { owner, repo, issue_number, body }: { owner: string, repo: string, issue_number: number, body: string }): Promise<IssueComment> {
	const params = { owner, repo, issue_number, body }
	const { data: comment } = await octokit.rest.issues.createComment(params)
	return comment
}

export async function createPullRequestReview(octokit: Octokit, { owner, repo, pull_number, body }: { owner: string, repo: string, pull_number: number, body: string }): Promise<PullRequestReview> {
	const params = { owner, repo, pull_number, body, event: 'COMMENT' as const }
	const { data: review } = await octokit.rest.pulls.createReview(params)
	return review
}
