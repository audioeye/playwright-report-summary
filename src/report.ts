import {
	JSONReport,
	JSONReportSpec,
	JSONReportSuite,
	JSONReportTest,
	JSONReportTestResult
} from '@playwright/test/reporter'
import { debug } from '@actions/core'

import { formatDuration, n, renderAccordion, upperCaseFirst } from './formatting'
import { icons, renderIcon } from './icons'

// interface Report {
// 	config: {
// 		configFile: string
// 		rootDir: string
// 		fullyParallel: boolean
// 		globalTimeout: number
// 		grep: Record<string, unknown>
// 		grepInvert: null
// 		maxFailures: number
// 		metadata: {
// 			actualWorkers: number
// 			totalTime: number
// 		}
// 		projects: Project[]
// 		shard?: {
// 			current: number
// 			total: number
// 		}
// 		version: string
// 		workers: number
// 	}
// 	suites: Suite[]
// 	errors: any[]
// }

// interface Project {
// 	id: string
// 	name: string
// }

// interface Suite {
// 	title: string
// 	file: string
// 	column: number
// 	line: number
// 	specs: Spec[]
// 	suites: Suite[]
// }

// interface Spec {
// 	title: string
// 	ok: boolean
// 	tags: string[]
// 	tests: Test[]
// 	id: string
// 	file: string
// 	line: number
// 	column: number
// }

// interface Test {
// 	timeout: number
// 	expectedStatus: 'passed' | 'skipped' | 'failed' | 'flaky' | string
// 	projectId: string
// 	projectName: string
// 	results: TestResult[]
// 	status: 'expected' | 'skipped' | string
// }

// interface TestResult {
// 	workerIndex: number
// 	status: 'passed' | 'failed' | 'skipped' | string
// 	duration: number
// 	error?: any
// 	errors: any[]
// 	retry: number
// 	startTime: string
// }

export interface ReportSummary {
	version: string
	started: Date
	duration: number
	workers: number
	shards: number
	projects: string[]
	files: string[]
	suites: SuiteSummary[]
	specs: SpecSummary[]
	tests: TestSummary[]
	failed: TestSummary[]
	passed: TestSummary[]
	flaky: TestSummary[]
	skipped: TestSummary[]
	results: TestResultSummary[]
}

interface SuiteSummary {
	file: string
	line: number
	column: number
	path: string[]
	title: string
	specs: SpecSummary[]
}

interface SpecSummary {
	ok: boolean
	file: string
	line: number
	column: number
	path: string[]
	title: string
	tests: TestSummary[]
}

interface TestSummary {
	passed: boolean
	failed: boolean
	flaky: boolean
	skipped: boolean
	file: string
	line: number
	column: number
	path: string[]
	title: string
	results: TestResultSummary[]
}

interface TestResultSummary {
	duration: number
	started: Date
}

interface ReportRenderOptions {
	commit?: string
	message?: string
	title?: string
	reportUrl?: string
	iconStyle?: keyof typeof icons
}

export function isValidReport(report: unknown): report is JSONReport {
	return report !== null && typeof report === 'object' && 'config' in report && 'errors' in report && 'suites' in report
}

export function parseReport(data: string): ReportSummary {
	const report: JSONReport = JSON.parse(data)
	if (!isValidReport(report)) {
		debug('Invalid report file')
		debug(data)
		throw new Error('Invalid JSON report file')
	}

	const { files, suites: allSuites } = extractSuiteInformation(report.suites)
	const suites: SuiteSummary[] = allSuites.map((suite) => parseSuite(suite), [] as SpecSummary[])
	const specs = suites.flatMap((suite) => suite.specs)
	const tests = specs.flatMap((spec) => spec.tests)
	const results = tests.flatMap((test) => test.results)
	const failed = tests.filter((test) => test.failed)
	const passed = tests.filter((test) => test.passed)
	const flaky = tests.filter((test) => test.flaky)
	const skipped = tests.filter((test) => test.skipped)

	const { duration, started } = getTotalDuration(report, results)
	const version: string = report.config.version
	const workers: number = report.config.metadata.actualWorkers || report.config.workers || 1
	const shards: number = report.config.shard?.total || 0
	const projects: string[] = report.config.projects.map((p) => p.name)

	return {
		version,
		started,
		duration,
		workers,
		shards,
		projects,
		files,
		suites,
		specs,
		tests,
		results,
		failed,
		passed,
		flaky,
		skipped
	}
}

function extractSuiteInformation(suites: JSONReportSuite[]): {
	files: string[],
	suites: JSONReportSuite[],
	specs: JSONReportSpec[]
} {
	let allFiles: string[] = [];
	let allSuites: JSONReportSuite[] = [];
	let allSpecs: JSONReportSpec[] = [];

	for (const suite of suites) {
		allFiles.push(suite.file);

		// Nested suites and their specs
		const { suites: nestedSuites, specs: nestedSpecs } = extractSuiteInformation(suite.suites ?? []);
		allSuites = allSuites.concat(nestedSuites);
		allSpecs = allSpecs.concat(nestedSpecs);

		// Current-suite specs
		allSpecs = allSpecs.concat(suite.specs ?? []);

		allSuites.push(suite);
	}

	return { files: allFiles, suites: allSuites, specs: allSpecs };
}

function parseSuite(suite: JSONReportSuite, parents: string[] = []): SuiteSummary {
	const { file, line, column } = suite
	const { title, path } = buildTitle(...parents, suite.title)
	const specs = suite.specs.map((spec) => parseSpec(spec, [...parents, suite.title]))
	return { file, line, column, path, title, specs }
}

function parseSpec(spec: JSONReportSpec, parents: string[] = []): SpecSummary {
	const { ok, file, line, column } = spec
	const { title, path } = buildTitle(...parents, spec.title)
	const tests = spec.tests.map((test) => parseTest(test, spec, parents))
	return { ok, file, line, column, path, title, tests }
}

function parseTest(test: JSONReportTest, spec: JSONReportSpec, parents: string[] = []): TestSummary {
	const { file, line, column } = spec
	const { status, projectName: project } = test
	const { title, path } = buildTitle(project, ...parents, spec.title)
	const results = test.results.map((result) => parseTestResult(result))
	const passed = status === 'expected'
	const failed = status === 'unexpected'
	const skipped = status === 'skipped'
	const flaky = status === 'flaky'
	return { passed, failed, flaky, skipped, results, title, path, file, line, column }
}

function parseTestResult({ duration, startTime }: JSONReportTestResult): TestResultSummary {
	return { duration, started: new Date(startTime) }
}

export function buildTitle(...paths: string[]): { title: string; path: string[] } {
	const path = paths.filter(Boolean)
	const title = path.join(' → ')
	return { title, path }
}

export function renderReportSummary(
	report: ReportSummary,
	{ commit, message, title, reportUrl, iconStyle }: ReportRenderOptions = {}
): string {
	const { duration, failed, passed, flaky, skipped } = report
	const icon = (symbol: string): string => renderIcon(symbol, { iconStyle })
	const paragraphs = []

	// Title

	paragraphs.push(`### ${title}`)

	// Passed/failed tests

	const tests = [
		failed.length ? `${icon('failed')}  **${failed.length} failed**` : ``,
		passed.length ? `${icon('passed')}  **${passed.length} passed**  ` : ``,
		flaky.length ? `${icon('flaky')}  **${flaky.length} flaky**  ` : ``,
		skipped.length ? `${icon('skipped')}  **${skipped.length} skipped**` : ``
	]
	paragraphs.push(tests.filter(Boolean).join('  \n'))

	// Stats about test run

	paragraphs.push(`#### Details`)

	const stats = [
		reportUrl ? `${icon('report')}  [Open report ↗︎](${reportUrl})` : '',
		`${icon('stats')}  ${report.tests.length} ${n('test', report.tests.length)} across ${report.suites.length} ${n(
			'suite',
			report.suites.length
		)}`,
		`${icon('duration')}  ${duration ? formatDuration(duration) : 'unknown'}`,
		commit && message ? `${icon('commit')}  ${message} (${commit.slice(0, 7)})` : '',
		commit && !message ? `${icon('commit')}  ${commit.slice(0, 7)}` : ''
	]
	paragraphs.push(stats.filter(Boolean).join('  \n'))

	// Lists of failed/skipped tests

	const listStatuses = ['failed', 'flaky', 'skipped'] as const
	const details = listStatuses.map((status) => {
		const tests = report[status]
		if (tests.length) {
			const summary = `${upperCaseFirst(status)} tests`
			const list = tests.map((test) => test.title).join('\n  ')
			const open = status === 'failed'
			return renderAccordion(summary, list, { open })
		}
	})
	paragraphs.push(
		details
			.filter(Boolean)
			.map((md) => (md as string).trim())
			.join('\n')
	)

	return paragraphs
		.map((p) => p.trim())
		.filter(Boolean)
		.join('\n\n')
}

function getTotalDuration(report: JSONReport, results: TestResultSummary[]): { duration: number; started: Date } {
	let duration = 0
	let started = new Date()
	const { totalTime } = report.config.metadata
	if (totalTime) {
		duration = totalTime
	} else {
		const sorted = results.sort((a, b) => a.started.getTime() - b.started.getTime())
		const first = sorted[0]
		const last = sorted[sorted.length - 1]
		if (first && last) {
			started = first.started
			duration = last.started.getTime() + last.duration - first.started.getTime()
		}
	}
	return { duration, started }
}
