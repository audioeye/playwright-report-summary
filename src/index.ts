import path from 'path'
import {
	getInput,
	getBooleanInput,
	setOutput,
	setFailed,
	startGroup,
	endGroup,
	debug,
	notice,
	warning,
	summary as setSummary
} from '@actions/core'
import { context, getOctokit } from '@actions/github'
import { fileExists, readFile } from './fs'
import { parseReport, renderReportSummary } from './report'

/**
 * The main function for the action.
 */
export async function run(): Promise<void> {
	try {
		await report()
	} catch (error) {
		if (error instanceof Error) {
			setFailed(error.message)
		}
	}
}

/**
 * Parse the Playwright report and post a comment on the PR.
 */
export async function report(): Promise<void> {
	const cwd = process.cwd()

	const { workflow, eventName, repo, payload } = context
	const { owner, number: pull_number } = context.issue

	const token = getInput('github-token')
	const reportFile = getInput('report-file', { required: true })
	const reportUrl = getInput('report-url')
	const reportTag = getInput('report-tag') || workflow
	const commentTitle = getInput('comment-title') || 'Playwright test results'
	const iconStyle = getInput('icon-style') || 'octicons'
	const jobSummary = getBooleanInput('job-summary')

	debug(`Report file: ${reportFile}`)
	debug(`Report url: ${reportUrl}`)
	debug(`Report tag: ${reportTag || '(none)'}`)
	debug(`Comment title: ${commentTitle}`)

	const base: { ref?: string; sha?: string } = {}
	const head: { ref?: string; sha?: string } = {}
	if (eventName === 'push') {
		base.ref = payload.ref
		base.sha = payload.before
		head.ref = payload.ref
		head.sha = payload.after
		console.log(`Commit pushed onto ${base.ref} (${head.sha})`)
	} else if (eventName === 'pull_request' || eventName === 'pull_request_target') {
		base.ref = payload.pull_request?.base?.ref
		base.sha = payload.pull_request?.base?.sha
		head.ref = payload.pull_request?.head?.ref
		head.sha = payload.pull_request?.head?.sha
		console.log(`PR #${pull_number} targeting ${base.ref} (${head.sha})`)
	} else {
		throw new Error(
			`Unsupported event type: ${eventName}. Only "pull_request", "pull_request_target", and "push" triggered workflows are currently supported.`
		)
	}

	const reportPath = path.resolve(cwd, reportFile)
	const reportExists = await fileExists(reportPath)
	if (!reportExists) {
		debug(`Failed to find report file at path ${reportPath}`)
		throw new Error(
			`Report file ${reportFile} not found. Make sure Playwright is configured to generate a JSON report.`
		)
	}

	const data = await readFile(reportPath)
	const report = parseReport(data)
	const summary = renderReportSummary(report, {
		commit: head.sha,
		title: commentTitle,
		reportUrl,
		iconStyle
	})

	const prefix = `<!-- playwright-report-github-action -- ${reportTag} -->`
	const body = `${prefix}\n\n${summary}`
	let commentId = null

	const octokit = getOctokit(token)

	if (eventName !== 'pull_request' && eventName !== 'pull_request_target') {
		console.log('No PR associated with this action run. Not posting a check or comment.')
	} else {
		startGroup(`Commenting test report on PR`)
		try {
			const { data: comments } = await octokit.rest.issues.listComments({
				...repo,
				issue_number: pull_number
			})
			const existingComment = comments.findLast((c) => c.body?.includes(prefix))
			commentId = existingComment?.id || null
		} catch (error: any) {
			console.error(`Error fetching existing comments: ${error.message}`)
		}

		if (commentId) {
			console.log(`Found previous comment #${commentId}`)
			try {
				await octokit.rest.issues.updateComment({
					...repo,
					comment_id: commentId,
					body
				})
				console.log(`Updated previous comment #${commentId}`)
			} catch (error: any) {
				console.error(`Error updating previous comment: ${error.message}`)
				commentId = null
			}
		}

		if (!commentId) {
			console.log('Creating new comment')
			try {
				const { data: newComment } = await octokit.rest.issues.createComment({
					...repo,
					issue_number: pull_number,
					body
				})
				commentId = newComment.id
				console.log(`Created new comment #${commentId}`)
			} catch (error: any) {
				console.error(`Error creating comment: ${error.message}`)
				console.log(`Submitting PR review comment instead...`)
				try {
					const { issue } = context
					await octokit.rest.pulls.createReview({
						owner,
						repo: issue.repo,
						pull_number: issue.number,
						event: 'COMMENT',
						body
					})
				} catch (error: any) {
					console.error(`Error creating PR review: ${error.message}`)
				}
			}
		}
		endGroup()
	}

	if (!commentId) {
		const intro = `Unable to comment on your PR â€” this can happen for PR's originating from a fork without write permissions. You can copy the test results directly into a comment using the markdown summary below:`
		warning(`${intro}\n\n${body}`, { title: 'Unable to comment on PR' })
	}

	if (jobSummary) {
		setSummary.addRaw(summary).write()
	}

	setOutput('summary', summary)
	setOutput('comment-id', commentId)
}

if (process.env.GITHUB_ACTIONS === 'true') {
	// eslint-disable-next-line @typescript-eslint/no-floating-promises
	run()
}
