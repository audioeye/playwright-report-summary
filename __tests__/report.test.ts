/**
 * Unit tests for src/report.ts
 */

import { expect } from '@jest/globals'
import { readFile } from '../src/fs'
import {
	ReportSummary,
	buildTitle,
	isValidReport,
	parseReport,
	parseReportFiles,
	parseReportSuites,
	renderReportSummary
} from '../src/report'

const defaultReport = 'report-valid.json'
const invalidReport = 'report-invalid.json'
const reportWithoutDuration = 'report-without-duration.json'
const shardedReport = 'report-sharded.json'
const nestedReport = 'report-nested.json'

async function getReport(file = defaultReport): Promise<string> {
	return await readFile(`__tests__/__fixtures__/${file}`)
}

async function getParsedReport(file = defaultReport): Promise<ReportSummary> {
	return parseReport(await getReport(file))
}

describe('isValidReport', () => {
	it('detects valid reports', async () => {
		const report = await getReport()
		expect(isValidReport(JSON.parse(report))).toBe(true)
	})
	it('detects invalid reports', async () => {
		const report = await getReport(invalidReport)
		expect(isValidReport([])).toBe(false)
		expect(isValidReport('')).toBe(false)
		expect(isValidReport(JSON.parse(report))).toBe(false)
	})
})

describe('buildTitle', () => {
	it('returns an object with path and title', async () => {
		const result = buildTitle('A', 'B')
		expect(result).toBeInstanceOf(Object)
		expect(result.path).toBeDefined()
		expect(result.title).toBeDefined()
	})
	it('concatenates and filters title segments', async () => {
		const { title } = buildTitle('A', 'B', '', 'C')
		expect(title).toBe('A › B › C')
	})
	it('concatenates and filters path segments', async () => {
		const { path } = buildTitle('A', '', 'B', 'C')
		expect(path).toStrictEqual(['A', 'B', 'C'])
	})
})

describe('parseReportFiles', () => {
	it('returns an array of root filenames', async () => {
		const report = JSON.parse(await getReport(nestedReport))
		const files = parseReportFiles(report)
		expect(files).toStrictEqual(['add.spec.ts', 'nested.spec.ts'])
	})
})

describe('parseReportSuites', () => {
	it('returns an array of root suite summaries', async () => {
		const report = JSON.parse(await getReport(nestedReport))
		const suites = parseReportSuites(report)
		expect(suites).toBeInstanceOf(Array)
		expect(suites.length).toBe(2)
		expect(suites[0].title).toBe('add.spec.ts')
	})
})

describe('parseReport', () => {
	it('returns an object', async () => {
		const parsed = await getParsedReport()
		expect(typeof parsed === 'object').toBe(true)
	})
	it('returns playwright version', async () => {
		const parsed = await getParsedReport()
		expect(parsed.version).toBe('1.37.1')
	})
	it('returns total duration', async () => {
		const parsed = await getParsedReport()
		expect(parsed.duration).toBe(1118.34)
	})
	it('calculates duration if missing', async () => {
		const parsed = await getParsedReport(reportWithoutDuration)
		expect(parsed.duration).toBe(943)
	})
	it('returns workers', async () => {
		const parsed = await getParsedReport()
		expect(parsed.workers).toBe(5)
	})
	it('returns shards', async () => {
		const parsed = await getParsedReport()
		expect(parsed.shards).toBe(2)
	})
	it('returns files', async () => {
		const parsed = await getParsedReport()
		expect(parsed.files.length).toBe(4)
	})
	it('returns suites', async () => {
		const parsed = await getParsedReport()
		expect(parsed.suites.length).toBe(4)
	})
	it('returns specs', async () => {
		const parsed = await getParsedReport()
		expect(parsed.specs.length).toBe(14)
	})
	it('counts tests', async () => {
		const parsed = await getParsedReport()
		expect(parsed.tests.length).toBe(14)
		expect(parsed.failed.length).toBe(2)
		expect(parsed.passed.length).toBe(10)
		expect(parsed.flaky.length).toBe(1)
		expect(parsed.skipped.length).toBe(1)
	})
	it('counts sharded tests', async () => {
		const parsed = await getParsedReport(shardedReport)
		expect(parsed.tests.length).toBe(27)
		expect(parsed.failed.length).toBe(1)
		expect(parsed.passed.length).toBe(22)
		expect(parsed.flaky.length).toBe(1)
		expect(parsed.skipped.length).toBe(3)
	})
	it('counts nested suites', async () => {
		const parsed = await getParsedReport(nestedReport)
		expect(parsed.suites.length).toBe(2)
		expect(parsed.specs.length).toBe(45)
		expect(parsed.tests.length).toBe(45)
	})
})

describe('renderReportSummary', () => {
	const renderOptions = {
		title: 'Test Report',
		reportUrl: 'https://example.com/report',
		additionalDetails: 'For more information, see our [documentation](https://example.com/docs)',
		commit: '1234567'
	}
	const getReportSummary = async (): Promise<string> =>
		renderReportSummary(parseReport(await getReport()), renderOptions)
	it('returns a string', async () => {
		const summary = await getReportSummary()
		expect(typeof summary === 'string').toBe(true)
	})
	it('matches snapshot', async () => {
		const summary = await getReportSummary()
		expect(summary).toMatchSnapshot()
	})
})
