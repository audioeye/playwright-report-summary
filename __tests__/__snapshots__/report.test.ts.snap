// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderReportSummary matches snapshot 1`] = `
"### Test Report

![failed](https://icongr.am/octicons/stop.svg?size=14&color=da3633)  **2 failed**  
![passed](https://icongr.am/octicons/check-circle.svg?size=14&color=3fb950)  **10 passed**    
![flaky](https://icongr.am/octicons/alert.svg?size=14&color=d29922)  **1 flaky**    
![skipped](https://icongr.am/octicons/skip.svg?size=14&color=0967d9)  **1 skipped**

#### Details

![report](https://icongr.am/octicons/package.svg?size=14&color=abb4bf)  [Open report ↗︎](https://example.com/report)  
![stats](https://icongr.am/octicons/pulse.svg?size=14&color=abb4bf)  14 tests across 4 suites  
![duration](https://icongr.am/octicons/clock.svg?size=14&color=abb4bf)  1.1 seconds  
![commit](https://icongr.am/octicons/git-pull-request.svg?size=14&color=abb4bf)  1234567  
![info](https://icongr.am/octicons/info.svg?size=14&color=abb4bf)  For more information, see our [documentation](https://example.com/docs)

<details open><summary><strong>Failed tests</strong></summary>

chromium › add.spec.ts › add › adds
  firefox › add.spec.ts › add › adds

</details>
<details ><summary><strong>Flaky tests</strong></summary>

chromium › multiply.spec.ts › multiply › multiplies

</details>
<details ><summary><strong>Skipped tests</strong></summary>

chromium › multiply.spec.ts › multiply › skips

</details>"
`;
